py/parse.c

```diff
@@ -1023,6 +1023,11 @@ STATIC void push_result_rule(parser_t *parser, size_t src_line, uint8_t rule_id,
 }
 
 mp_parse_tree_t mp_parse(mp_lexer_t *lex, mp_parse_input_kind_t input_kind) {
+    // Set exception handler to free the lexer if an exception is raised.
+    nlr_jump_callback_node_call_function_1_t ctx;
+    ctx.func = mp_lexer_free;
+    ctx.arg = lex;
```

Comment:
Instead of changing the signature of `mp_lexer_free` how about creating a macro...

```c
#define MP_DEFINE_NLR_CALLBACK_FUNCTION_1(ctx, f, d) \
nlr_jump_callback_node_call_function_1_t ctx = { \
    .func = (void (*)(void *))(f), \
    .arg = (void *)(d), \
}
```

Then usage is also less verbose.
```suggestion
    MP_DEFINE_NLR_CALLBACK_FUNCTION_1(ctx, eader->close, reader->data);
```

Reply:
Yeah, happy to make this change.

The reason I changed the signature was because I didn't want to risk any edge-case C incompatibility when casting a function to a different function pointer.  I'm actually not sure of the rules here, do you know what is legal/allowed by the C standard with respect to casting function pointers?

Reply:
I'm not aware of any compiler restrictions on functions pointers (e.g. https://stackoverflow.com/a/66206802/1976323). And I've seen plenty of code (including CPython extensions) that make casts like this. So I don't think there is anything to worry about here. We do lose some type safety, but it doesn't seem like something that would be easy to mess up in the first place.

Reply:
I made this change.